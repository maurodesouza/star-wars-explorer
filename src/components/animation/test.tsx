import { screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import { Animation, AnimationProps } from '.';

jest.mock('react-lottie', () => ({
  __esModule: true,
  default: function Mock() {
    return <svg data-testid="lottie" />;
  },
}));

const dataMock: AnimationProps = {
  animation: 'loading',
  label: 'foo',
};

describe('<Animation />', () => {
  it('should render the animation styles correct', () => {
    const { container } = renderWithProviders(<Animation {...dataMock} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the animation component correct', () => {
    renderWithProviders(<Animation {...dataMock} />);

    const label = screen.getByText(dataMock.label!);
    const lottie = screen.getByTestId('lottie');

    expect(label).toBeInTheDocument();
    expect(lottie).toBeInTheDocument();
  });
});
