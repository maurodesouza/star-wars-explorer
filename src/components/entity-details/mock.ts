import { Entities } from 'types';
import { getImageUrl } from 'utils';

const mock = {
  extras: {
    key: 'value',
    other: 'foo',
    bar: 'bla',
    foo: 'here',
  },

  image: getImageUrl(Entities.CHARACTERS, '2'),
};

export { mock };
