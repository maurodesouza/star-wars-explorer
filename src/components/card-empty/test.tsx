import { screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import { EmptyCard } from '.';

const dataMock = {
  id: 'test',
};

describe('<EmptyCard />', () => {
  it('should render the empty card styles correct', () => {
    const { container } = renderWithProviders(<EmptyCard {...dataMock} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the empty card component correct', () => {
    renderWithProviders(<EmptyCard {...dataMock} />);

    const label = screen.getByText(`${dataMock.id}#`);

    expect(label).toBeInTheDocument();
  });
});
