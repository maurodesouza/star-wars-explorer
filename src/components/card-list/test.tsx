import { cleanup } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import { Entities } from 'types';
import { createArray } from 'utils';

import { CardList } from '.';

jest.mock('components', () => ({
  __esModule: true,
  Card: function Mock() {
    return <div data-testid="card" />;
  },
}));

const itemMock = {
  id: '1',
  title: 'C-3PO',
  image: 'image',
  entity: Entities.CHARACTERS,
};

describe('<CardList />', () => {
  describe('styles', () => {
    beforeAll(() => {
      jest.resetAllMocks();
    });

    it('should render the card list styles correct', () => {
      const { container } = renderWithProviders(<CardList />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  it('should render the items card correct', () => {
    const counts = [4, 8];

    counts.forEach(count => {
      const items = createArray(count).map((_, index) => ({
        ...itemMock,
        id: String(index),
      }));

      const render = renderWithProviders(<CardList items={items} />);
      const findedItems = render.getAllByTestId(/card/i);

      expect(findedItems.length).toBe(count);

      cleanup();
    });
  });

  it('should render the fills correct', () => {
    const minCountToFill = 4;
    const counts = [1, 3, 4, 6];

    counts.forEach(count => {
      const result = count - minCountToFill;
      const lackForFill = result <= 0 ? Math.abs(result) : 0;

      const items = createArray(count).map((_, index) => ({
        ...itemMock,
        id: String(index),
      }));

      const { container } = renderWithProviders(<CardList items={items} />);
      const findedItems = container.firstElementChild!.querySelectorAll(
        'div:not([data-testId]'
      );

      expect(findedItems.length).toBe(lackForFill);
      cleanup();
    });
  });
});
