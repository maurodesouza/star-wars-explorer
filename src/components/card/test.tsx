/* eslint-disable @typescript-eslint/no-var-requires */

import { act, fireEvent, screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import { Entities } from 'types';
import { Card } from '.';

const useRouter = jest.spyOn(require('next/router'), 'useRouter');
const push = jest.fn();

useRouter.mockImplementation(() => ({
  push,
}));

const mockData = {
  id: '1',
  label: 'C-3PO',
  image:
    'https://raw.githubusercontent.com/tbone849/star-wars-guide/master/build/assets/img/characters/2.jpg',
};

describe('<Card />', () => {
  it('should render the card correct', () => {
    renderWithProviders(<Card {...mockData} />);

    expect(
      screen.getByRole('img', { name: mockData.label })
    ).toBeInTheDocument();

    expect(
      screen.getByRole('heading', { name: mockData.label })
    ).toBeInTheDocument();
  });

  it("should render the no image found element when can't load image url", async () => {
    renderWithProviders(<Card {...mockData} />);

    const image = screen.getByRole('img', { name: mockData.label });

    act(() => {
      fireEvent.error(image);
    });

    const noImageFoundElement = screen.getByText(/no found/i);

    expect(noImageFoundElement).toBeInTheDocument();
  });

  it('should render the card styles correct', () => {
    const { container } = renderWithProviders(<Card {...mockData} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should navigate to entity page', () => {
    const { container } = renderWithProviders(<Card {...mockData} />);

    fireEvent.click(container.firstChild!);

    expect(push).toBeCalledTimes(1);
    expect(push).toBeCalledWith(`/${Entities.CHARACTERS}/${mockData.label}`);
  });
});
