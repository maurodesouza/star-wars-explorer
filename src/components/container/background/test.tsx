import { renderWithProviders } from 'utils/test/helpers';
import { Background } from '.';

describe('components/container', () => {
  describe('<Background />', () => {
    it('should render the background styles correct', () => {
      const { container } = renderWithProviders(<Background />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
