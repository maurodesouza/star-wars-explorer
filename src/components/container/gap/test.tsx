import { renderWithProviders } from 'utils/test/helpers';
import { Gap } from '.';

describe('components/container', () => {
  describe('<Gap />', () => {
    it('should render the gap styles correct', () => {
      const { container } = renderWithProviders(<Gap />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
