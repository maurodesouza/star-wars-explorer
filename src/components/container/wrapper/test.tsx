import { renderWithProviders } from 'utils/test/helpers';
import { Wrapper } from '.';

describe('components/container', () => {
  describe('<Wrapper />', () => {
    it('should render the wrapper styles correct', () => {
      const { container } = renderWithProviders(<Wrapper />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
