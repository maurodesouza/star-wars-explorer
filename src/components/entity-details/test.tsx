import { screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import { Entities } from 'types';

import { mock } from './mock';
import { EntityDetails, EntityDetailsProps } from '.';

const dataMock: EntityDetailsProps = {
  ...mock,
  id: '1',
  title: 'label',
  entity: Entities.CHARACTERS,
};

describe('<EntityDetails />', () => {
  it('should render the entity details styles correct', () => {
    const { container } = renderWithProviders(<EntityDetails {...dataMock} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the entity details component correct', () => {
    renderWithProviders(<EntityDetails {...dataMock} />);

    const image = screen.getByRole('img', { name: dataMock.title });

    expect(image).toBeInTheDocument();

    Object.entries(mock.extras)
      .reduce((arr, [key, value]) => [...arr, `${key}:`, value], [] as string[])
      .forEach(text => {
        const el = screen.getByText(text);

        expect(el).toBeInTheDocument();
      });
  });

  it('should no render content if no extras was passed', () => {
    const { container } = renderWithProviders(
      <EntityDetails {...dataMock} extras={undefined} />
    );

    const component = container.children[0];
    const content = component.children[1];

    expect(content.children.length).toBe(0);
  });
});
