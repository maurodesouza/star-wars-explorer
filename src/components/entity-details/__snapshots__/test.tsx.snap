// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EntityDetails /> should render the entity details styles correct 1`] = `
.c4 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c3 {
  width: 2.4rem;
  height: 2.4rem;
  color: #E62429;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  border: 0;
  background: none;
  cursor: pointer;
}

.c3 * {
  cursor: pointer;
}

.c0 {
  margin-top: 4.0rem;
  gap: 2.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
}

.c1 {
  position: relative;
  width: min(100%,30rem);
  aspect-ratio: 0.65;
  border-radius: 0.8rem;
  background: #FFF;
  overflow: hidden;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  border: 1px solid #FFF;
}

.c2 {
  position: absolute;
  bottom: 1.6rem;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.8rem;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.8rem;
  font-size: 1.6rem;
}

.c7 {
  text-transform: capitalize;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

@media (min-width:768px) {
  .c0 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

@media (min-width:768px) {
  .c1 {
    -webkit-align-self: unset;
    -ms-flex-item-align: unset;
    align-self: unset;
    width: 20rem;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <img
      alt="label"
      layout="fill"
      src="https://raw.githubusercontent.com/tbone849/star-wars-guide/master/build/assets/img/characters/2.jpg"
      style="position: absolute; inset: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
    />
    <div
      class="c2"
    >
      <button
        class="c3"
      >
        <svg
          aria-hidden="true"
          aria-label="Add to favorites"
          class="c4"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55l-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    class="c5"
  >
    <div
      class="c6"
    >
      <strong
        class="c7"
      >
        key
        :
      </strong>
      <span
        class=""
      >
        value
      </span>
    </div>
    <div
      class="c6"
    >
      <strong
        class="c7"
      >
        other
        :
      </strong>
      <span
        class=""
      >
        foo
      </span>
    </div>
    <div
      class="c6"
    >
      <strong
        class="c7"
      >
        bar
        :
      </strong>
      <span
        class=""
      >
        bla
      </span>
    </div>
    <div
      class="c6"
    >
      <strong
        class="c7"
      >
        foo
        :
      </strong>
      <span
        class=""
      >
        here
      </span>
    </div>
  </div>
</div>
`;
